[PyTorch] Avoid dispatch registration for static dispatch

larryliu0820 opened the pull request:
Summary: For static dispatch, we don't need to register dispatch rules into dispatcher any more, since the C++ API will call corresponding native function directly.

Test Plan:  `gen.py -d cg/torch/share/ATen --static_dispatch_backend=CPU `
Generated `RegisterCPU.cpp`:
```

TORCH_LIBRARY_IMPL(aten, CPU, m) {

}
```
Need to find a way to test this in current static dispatch use case.

Differential Revision: D33670265

bdhirsh commented on aten/src/ATen/templates/RegisterDispatchKey.cpp:
```diff
@@ -57,9 +57,9 @@ namespace {
-}
+${dispatch_registrations_begin} // TORCH_LIBRARY_IMPL(aten, ${DispatchKey}, m) {
+${dispatch_registrations}
+${dispatch_registrations_end} // }
```
can we get away with leaving these in and just passing in an empty list of registrations?

larryliu0820 replied:
Are you suggesting that we could leave something like:
```
TORCH_LIBRARY_IMPL(aten, CPU, m) {

}
```
I'm not sure what's the underlying effect though

larryliu0820 replied:
Also @bdhirsh can you please help tag whoever knows how to test this in static dispatch use case?

bdhirsh replied:
yep. That should be a no-op, so equivalent. And it makes the codegen a little cleaner, since we don't have to add as much special handling across the file for static dispatch.

bdhirsh replied:
> can you please help tag whoever knows how to test this in static dispatch use case?

Right. Hey @swolchok, the mobile team is planning on making a few tweaks to static dispatch, that we don't expect to impact the server use case. Do you think we can rely on the existing CI (and internal tests, if there are any) to catch issues? Mostly just wondering if you think it's worth running any manual internal tests.

This specific PR updates the static dispatch codegen to avoid registering anything to the dispatcher (seems harmless, since static dispatch doesn't use the dispatcher).

larryliu0820 commented:
Temporarily disable `linux-xenial-py3-clang5-mobile-custom-build-static`, will re-enable in the following PR.

larryliu0820 committed:
[PyTorch] Avoid dispatch registration for static dispatch (#71504)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71504

For static dispatch, we don't need to register dispatch rules into dispatcher any more, since the C++ API will call corresponding native function directly.

Test Plan:
```
gen.py -d cg/torch/share/ATen --static_dispatch_backend=CPU
```

Generated `RegisterCPU.cpp`:
```

TORCH_LIBRARY_IMPL(aten, CPU, m) {

}
```
Need to find a way to test this in current static dispatch use case.

Differential Revision: D33670265

fbshipit-source-id: 81b87d0fe5978d5c82e6c37f0c3d9b6a76be6780

larryliu0820 commented:
Absorbed by #71782.
